src/Security/Authentication/Certificate/src/CertificateAuthenticationHandler.cs

```diff
@@ -0,0 +1,269 @@
+            }
+
+            return chainPolicy;
+        }
+
+        private ClaimsPrincipal CreatePrincipal(X509Certificate2 certificate)
+        {
+            var claims = new List<Claim>();
+
+            var issuer = certificate.Issuer;
+            claims.Add(new Claim("issuer", issuer, ClaimValueTypes.String, Options.ClaimsIssuer));
+
+            var thumbprint = certificate.Thumbprint;
+            claims.Add(new Claim(ClaimTypes.Thumbprint, thumbprint, ClaimValueTypes.Base64Binary, Options.ClaimsIssuer));
+
+            var value = certificate.SubjectName.Name;
+            if (!string.IsNullOrWhiteSpace(value))
+            {
+                claims.Add(new Claim(ClaimTypes.X500DistinguishedName, value, ClaimValueTypes.String, Options.ClaimsIssuer));
+            }
+
+            value = certificate.SerialNumber;
+            if (!string.IsNullOrWhiteSpace(value))
+            {
+                claims.Add(new Claim(ClaimTypes.SerialNumber, value, ClaimValueTypes.String, Options.ClaimsIssuer));
+            }
+
+            value = certificate.GetNameInfo(X509NameType.DnsName, false);
+            if (!string.IsNullOrWhiteSpace(value))
+            {
+                claims.Add(new Claim(ClaimTypes.Dns, value, ClaimValueTypes.String, Options.ClaimsIssuer));
+            }
+
+            value = certificate.GetNameInfo(X509NameType.SimpleName, false);
+            if (!string.IsNullOrWhiteSpace(value))
+            {
+                claims.Add(new Claim(ClaimTypes.Name, value, ClaimValueTypes.String, Options.ClaimsIssuer));
+            }
+
+            value = certificate.GetNameInfo(X509NameType.EmailName, false);
+            if (!string.IsNullOrWhiteSpace(value))
+            {
+                claims.Add(new Claim(ClaimTypes.Email, value, ClaimValueTypes.String, Options.ClaimsIssuer));
+            }
+
+            value = certificate.GetNameInfo(X509NameType.UpnName, false);
+            if (!string.IsNullOrWhiteSpace(value))
+            {
+                claims.Add(new Claim(ClaimTypes.Upn, value, ClaimValueTypes.String, Options.ClaimsIssuer));
+            }
+
+            value = certificate.GetNameInfo(X509NameType.UrlName, false);
+            if (!string.IsNullOrWhiteSpace(value))
+            {
+                claims.Add(new Claim(ClaimTypes.Uri, value, ClaimValueTypes.String, Options.ClaimsIssuer));
+            }
+
+            var identity = new ClaimsIdentity(claims, CertificateAuthenticationDefaults.AuthenticationScheme);
+            return new ClaimsPrincipal(identity);
+        }
+
+        private AuthenticateResult Success(ClaimsPrincipal principal, X509Certificate2 certificate)
+        {
+            var props = new AuthenticationProperties
+            {
+                Items =
+                {
+                    {
+                        CertificateAuthenticationDefaults.CertificateItemsKey, certificate.GetRawCertDataString()
```

Comment:
Why would this get put this in the properties collection? Use the ClaimsIdentity.BootstrapContext if you must. Whatever you do, do it before the ValidateCertificate event.

They could also just read the cert off the Connection if they wanted, not sure why we need to bundle it.

Reply:
@blowdart any issue with just removing this from the items?

Reply:
@leastprivilege - this was from you, so, does it need to be a claim?

Reply:
Removed?